File: M3Tools/Dialogs/Listeners/EditListenerDialog.vb
Imports System.Windows.Forms

Namespace Dialogs
	Public Class EditListenerDialog
		Private Event ListenerChanged()
		'Private _newInfo As Types.Listener

		Public Property Original As Types.Listener

		Public ReadOnly Property Listener As Types.Listener
			Get
				Return New Types.Listener(Original.Id, ListenerName, ListenerEmail)
			End Get
		End Property

		Public Property ListenerName As String
			Get
				Return gi_Name.Text
			End Get
			Set(value As String)
				gi_Name.Text = value
			End Set
		End Property

		Public Property ListenerEmail As String
			Get
				Return gi_Email.Text
			End Get
			Set(value As String)
				gi_Email.Text = value
			End Set
		End Property

		Sub New(listener As Types.Listener)
			' This call is required by the designer.
			InitializeComponent()

			' Add any initialization after the InitializeComponent() call.
			Me.Original = listener
			RaiseEvent ListenerChanged()
		End Sub

		Private Sub FinishDialog(sender As Object, e As EventArgs) Handles btn_Save.Click
			If Not ChangeDetected() Then
				Return
			End If

			Me.DialogResult = DialogResult.OK
			Me.Close()
		End Sub

		Private Sub CancelDialog(sender As Object, e As EventArgs) Handles btn_Cancel.Click
			Me.DialogResult = DialogResult.Cancel
			Me.Close()
		End Sub

		Private Sub ListenerUpdated() Handles Me.ListenerChanged
			ListenerName = Original.Name
			ListenerEmail = Original.Email
		End Sub

		Private Function ChangeDetected() As Boolean
			If ListenerName <> Original.Name Then
				Return True
			End If

			If ListenerEmail <> Original.Email Then
				Return True
			End If

			Return False
		End Function
	End Class
End Namespace

File: M3Tools/Dialogs/Listeners/EditListenerDialog.cs
using System;
using System.Windows.Forms;

// TODO: Create a base class for these edit dialogs and other components I use for easier updating

namespace SPPBC.M3Tools.Dialogs
{
	/// <summary>
	/// A dialog for editing the info of a listener
	/// </summary>
	public partial class EditListenerDialog
	{
		private Types.Listener _original;
		private event EventHandler ListenerChanged;

		/// <summary>
		/// The original info for the listener
		/// </summary>
		public Types.Listener Original
		{
			get => _original;
			private set
			{
				_original = value;
				ListenerName = _original.Name;
				ListenerEmail = _original.Email;
				Text = string.Format(Text, _original.Name);
			}
		}

		/// <summary>
		/// The current listener info
		/// </summary>
		public Types.Listener Listener => new(Original.Id, ListenerName, ListenerEmail);

		/// <summary>
		/// The name of the listener
		/// </summary>
		public string ListenerName
		{
			get => gi_Name.Text;
			private set => gi_Name.Text = value;
		}

		/// <summary>
		/// The email of the listener
		/// </summary>
		public string ListenerEmail
		{
			get => gi_Email.Text;
			private set => gi_Email.Text = value;
		}

		private EditListenerDialog()
		{
			InitializeComponent();
		}

		/// <summary>
		/// <inheritdoc/>
		/// </summary>
		/// <param name="listener"></param>
		public EditListenerDialog(Types.Listener listener) : this()
		{
			// Add any initialization after the InitializeComponent() call.
			Original = listener;
		}

		private void FinishDialog(object sender, EventArgs e)
		{
			if (Listener == Original)
			{
				return;
			}

			DialogResult = DialogResult.OK;
			Close();
		}

		private void CancelDialog(object sender, EventArgs e)
		{
			DialogResult = DialogResult.Cancel;
			Close();
		}
	}
}
